util/Pos/Util/Log/LogSafe.hs

```diff
@@ -0,0 +1,397 @@
+       ) where
+
+-- Universum has its own Rube Goldberg variant of 'Foldable' which we do not
+-- want. It would be great if we could write
+--   import           Universum hiding (toList, foldMap)
+-- but HLint insists that this is not OK because toList and foldMap are never
+-- used unqualified. The hiding in fact makes it clearer for the human reader
+-- what's going on.
+import           Universum
+
+import           Control.Concurrent (myThreadId)
+import           Control.Lens (each)
+import           Control.Monad.Trans (MonadTrans)
+import           Data.Foldable (length, null)
+import           Data.Map.Strict (lookup)
+import           Data.Reflection (Reifies (..), reify)
+import           Data.Text.Lazy.Builder (Builder)
+import           Formatting (bprint, build, fconst, later, mapf, (%))
+import qualified Formatting.Buildable
+import           Formatting.Internal (Format (..))
+import qualified Language.Haskell.TH as TH
+
+import           Pos.Util.Log (CanLog (..), LogContext, LoggingHandler,
+                     Severity (..), WithLogger)
+import           Pos.Util.Log.Internal (getConfig, getLogEnv, sev2klog)
+import           Pos.Util.LoggerConfig (LogHandler (..), LogSecurityLevel (..),
+                     lcLoggerTree, lhName, ltHandlers)
+
+import qualified Katip as K
+import qualified Katip.Core as KC
+
+----------------------------------------------------------------------------
+-- Logging
+----------------------------------------------------------------------------
+
+-- | Modifies logging so that it writes only to files specified by @s@ type.
+newtype SelectiveLogWrapped s m a = SelectiveLogWrapped
+    { getSecureLogWrapped :: m a
+    } deriving (Functor, Applicative, Monad, MonadIO)
+
+instance MonadTrans (SelectiveLogWrapped s) where
+    lift = SelectiveLogWrapped
+
+-- | Whether to log to given log handler.
+type SelectionMode = LogSecurityLevel -> Bool
+
+selectPublicLogs :: SelectionMode
+selectPublicLogs = \case
+    PublicLogLevel -> True
+    _              -> False
+
+selectSecretLogs :: SelectionMode
+selectSecretLogs = not . selectPublicLogs
+
+-- | conditionally log a message to all scribes if they match 'SelectionMode'
+logMCond :: (LogContext m) => LoggingHandler -> Severity -> Text -> SelectionMode -> m ()
+logMCond lh sev msg cond = do
+    ctx <- K.getKatipContext
+    ns  <- K.getKatipNamespace
+    logItemS lh ctx ns Nothing (sev2klog sev) cond $ K.logStr msg
+
+-- | this emulates katip's 'logItem' function, but only outputs the message
+--   to scribes which match the 'SelectionMode'
+logItemS
+    :: (K.LogItem a, K.Katip m)
+    => LoggingHandler
+    -> a
+    -> K.Namespace
+    -> Maybe TH.Loc
+    -> K.Severity
+    -> SelectionMode
+    -> K.LogStr
+    -> m ()
+logItemS lhandler a ns loc sev cond msg = do
+    mayle <- liftIO $ getLogEnv lhandler
+    case mayle of
+        Nothing              -> error "logging not yet initialized. Abort."
+        Just le@K.LogEnv{..} -> do
```

Comment:
I am not a fan of the RecordWildcard syntax extension but it is very commonly used in our code base so I usually let it ride. My objections to RWC is minor when the data constructor is in the same file and the objection grows the further away the use case is from the definition; eg from minor to major objection, same file, same package, same git repository.

In this case `K.LogEnv` is in an upstream package and I would need to got to Hackage to find the details of that data constructor. This is really not very friendly to the code *reader*. In this case using the field accessor makes the code some much easier to read and it becomes obvious where values like `_logEnvApp` comes from.

Reply:
true, too lazy! Will replace it with the actual field.
Explicitly named fields are easier to follow; also for tooling (i.e. tags).